File: TPS/TP1/arquivos/arquivo.c
#include <stdio.h>

void gravarValoresArquivo(FILE *arquivo, int n) {
    double numero;
    for (int i = 0; i < n; i++) {
        scanf("%lf", &numero);  
        fwrite(&numero, sizeof(double), 1, arquivo);  
    }
}

void lerValoresReverso(FILE *arquivo, int n) {
    double numero;
    for (int i = n - 1; i >= 0; i--) {
        fseek(arquivo, i * sizeof(double), SEEK_SET);  
        fread(&numero, sizeof(double), 1, arquivo);
        printf("%.3lf\n", numero);  
    }
}

int main() {
    int n;

    scanf("%d", &n);

    while (n != 200) {

        FILE *arquivo = fopen("valores.txt", "wb");
        if (arquivo == NULL) {
            printf("Erro ao abrir o arquivo.\n");
            return 1;
        }

        gravarValoresArquivo(arquivo, n);

        fclose(arquivo);

        arquivo = fopen("valores.txt", "rb");
        if (arquivo == NULL) {
            printf("Erro ao abrir o arquivo para leitura.\n");
            return 1;
        }

        lerValoresReverso(arquivo, n);

      
        fclose(arquivo);

        
        scanf("%d", &n);
    }

    return 0;
}

File: TPS/TP1/arquivos/Arquivo.java
//package TP1.arquivos;

import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class Arquivo {

    public static boolean verifica(int n){
        if (n != 200) {
            return false;
        }
         return true;
    }

    public static void main(String[] args) {
        Scanner ler = new Scanner(System.in);
        
        int n = ler.nextInt();
        while (verifica(n)) {

            gravarValoresArquivo(n, ler);
         
            lerValoresReverso(n);
            
            n = ler.nextInt();
        }

        ler.close();
    }

    public static void gravarValoresArquivo(int n, Scanner scanner) {
        RandomAccessFile arquivo = null;

        try {
            arquivo = new RandomAccessFile("valores.txt", "rw");

            for (int i = 0; i < n; i++) {
                double numero = scanner.nextDouble();
                arquivo.writeDouble(numero);
            }

        } catch (IOException e) {
            System.out.println("Erro ao gravar os números no arquivo.");
        } finally {
            fecharArquivo(arquivo);
        }
    }

    public static void lerValoresReverso(int n) {
        RandomAccessFile arquivo = null;

        try {
            arquivo = new RandomAccessFile("valores.txt", "r");

            for (int i = n - 1; i >= 0; i--) {
                arquivo.seek(i * 8); 
                double numeroLido = arquivo.readDouble();
                System.out.println(numeroLido);
            }

        } catch (IOException e) {
            System.out.println("Erro ao ler os números do arquivo.");
        } finally {
            fecharArquivo(arquivo);
        }
    }

    public static void fecharArquivo(RandomAccessFile arquivo) {
        try {
            if (arquivo != null) {
                arquivo.close();
            }
        } catch (IOException e) {
            System.out.println("Erro ao fechar o arquivo.");
        }
    }
}